src/ImageRunModal.jsx

```diff
@@ -27,13 +27,112 @@ const units = {
     },
 };
 
+class PublishPortsForm extends React.Component {
+    constructor(props) {
+        super(props);
+        this.state = {
+            publish: [{ containerPort: null, hostPort: null, protocol: 'TCP' }],
+        };
+        this.removePort = this.removePort.bind(this);
+        this.addPort = this.addPort.bind(this);
+        this.onPortChange = this.onPortChange.bind(this);
+        this.validateAndChange = this.validateAndChange.bind(this);
+    }
+
+    removePort(idx, field, value) {
+        let ports = this.state.publish;
+        ports.splice(idx, 1);
```

Comment:
You must not modify state in any way other than `setState()`, and this does (through aliasing). Make a copy with `this.state.publish.concat()`.